### PR TITLE
working on create vocab

### PR DESCRIPTION
**Pull request by zikowang**
| Status  | Type  |
| :---: | :---: |
| :white_check_mark: Ready | Feature |

## Description

Create new Vocab for selected package and group through add vocab screen
## Motivation and Context
- [x] User can add new vocabulary
- [x] User can add new language package while in add vocabulary process
- [x] User can add new Group to an Specific language package while in add vocabulary process

## How Has This Been Tested?
locally

## Screenshots / GIFs (if appropriate):

Apr-07-2021 15-20-42
Bildschirmfoto 2021-04-16 um 21 56 09
Bildschirmfoto 2021-04-16 um 21 56 12
Bildschirmfoto 2021-04-16 um 21 56 46
Bildschirmfoto 2021-04-16 um 21 56 52

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (non-breaking change which cleans up / improves existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the **CONTRIBUTING** document.
- [x] I have considered the accessibility of my changes (i.e. did I add proper content descriptions to images, or run my changes with talkback enabled?)
- [x] I have documented my code if needed

### CODE REVIEW

File: src/Components/GroupForm/GroupForm.jsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  const fetchPackages = useCallback(() => {
+    getPackages(true)
+      .then(({ data }) => {
+        setLanguagePackages(data);
+      })
+      .catch(function (err) {});
+  }, []);
+
+  //make api call to add vocab package
+  async function submitHandler() {
```
Comment (luwol03) [inline src/Components/GroupForm/GroupForm.jsx:35]: use arrow function (please search thourgh the complete PR if you used normal functions somewhere else.)

File: src/Components/Snackbar/Snackbar.scss [hunk 2]
```diff
@@ -0,0 +1,45 @@
+@import "../../colors";
+
+.snackbar {
+  position: absolute;
+  bottom: 15px;
+  opacity: 0;
+  // left: calc(50% + 32px);
```
Comment (luwol03) [inline src/Components/Snackbar/Snackbar.scss:7]: could you please remove this comments?

File: src/context/SnackbarContext.jsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    },
+    [t]
+  );
+
+  useEffect(() => {
+    if (!snack.show) {
+      return;
+    }
+
+    let timer = setTimeout(() => {
```
Comment (luwol03) [inline src/context/SnackbarContext.jsx:31]: could we use const instead?

File: src/Components/GroupForm/GroupForm.jsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    const newGroup = {
+      name,
+      active: true,
```
Comment (luwol03) [inline src/Components/GroupForm/GroupForm.jsx:38]: can we add a checkbox for the active state?

File: src/Components/GroupForm/GroupForm.jsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      })
+    );
+  }, [languagePackages]);
+
+  return (
+    <div className="group-form">
+      <div className="form-wrapper">
+        <div className="dropdown">
+          <div className="select-wrapper">
+            <Select
```
Comment (luwol03) [inline src/Components/GroupForm/GroupForm.jsx:86]: can we use the custom select option also here?

File: src/screens/AddVocab/AddVocab.jsx [hunk 6]
```diff
@@ -1,198 +1,285 @@
...
+              onChange={(v) => {
+                setSelectedGroup(v);
+              }}
+              value={selectedGroup}
+              noOptionsMessage={t("screens.addVocab.noGroupsMessage")}
+            />
+          </div>
+        </div>
+        <div className="input-fields">
+          <TextInput
```
Comment (luwol03) [inline src/screens/AddVocab/AddVocab.jsx:231]: we also need a active checkbox and an description field. See: vocascan/vocascan-server#32

**Review by luwol03 — CHANGES_REQUESTED** [review]

File: src/Components/GroupForm/GroupForm.jsx (hunk 5 above)
Comment (zikowang) [reply src/Components/GroupForm/GroupForm.jsx:86]: I think we do use the custom select here. Do I miss something?

File: src/Components/GroupForm/GroupForm.jsx (hunk 4 above)
Comment (zikowang) [reply src/Components/GroupForm/GroupForm.jsx]: I would like to use the switcher for the active state.

File: src/screens/AddVocab/AddVocab.jsx (hunk 6 above)
Comment (zikowang) [reply src/screens/AddVocab/AddVocab.jsx:247]: I would like to use the switcher for the active state.

File: src/Components/GroupForm/GroupForm.jsx (hunk 5 above)
Comment (luwol03) [reply src/Components/GroupForm/GroupForm.jsx:97]: I mean the select with the flags on the right side.

File: src/screens/AddVocab/AddVocab.jsx [hunk 7]
```diff
@@ -1,198 +1,315 @@
...
+        description,
+      },
+      active
+    )
+      .then((response) => {
+        onClear();
+        showSnack("success", t("screens.addVocab.saveSuccessMessage"));
+      })
+      .catch((e) => {
+        showSnack("error", t("screens.addVocab.saveSuccessMessage"));
```
Comment (luwol03) [inline src/screens/AddVocab/AddVocab.jsx:123]: why is this a success message?

File: src/screens/AddVocab/AddVocab.jsx [hunk 8]
```diff
@@ -1,198 +1,315 @@
...
+            placeholder={t("global.description")}
             onChange={(value) => {
-              setGroupName(value);
+              setDescription(value);
             }}
-            value={groupName}
-            noOptionsMessage={t("screens.addVocab.noGroupsMessage")}
+            value={description}
+          />
+          <Switch
```
Comment (luwol03) [inline src/screens/AddVocab/AddVocab.jsx:272]: could you try to make the switch a bit smaller?

File: src/screens/AddVocab/AddVocab.jsx [hunk 9]
```diff
@@ -1,198 +1,315 @@
...
+
+    createVocabulary(
+      selectedPackage.value,
+      selectedGroup.value,
+      {
+        name: foreignWord,
+        translations: submitTranslations,
+        description,
+      },
+      active
```
Comment (luwol03) [inline src/screens/AddVocab/AddVocab.jsx:116]: For now there are two disabled states. So we need two switches. One for disable the whole vocab and one for moving the vocab into a special drawer called staged.

File: src/screens/AddVocab/AddVocab.jsx (hunk 7 above)
Comment (zikowang) [reply src/screens/AddVocab/AddVocab.jsx]: my bad

File: src/screens/AddVocab/AddVocab.jsx (hunk 8 above)
Comment (zikowang) [reply src/screens/AddVocab/AddVocab.jsx:275]: done

File: src/Components/Tooltip/Tooltip.jsx [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import React from "react";
+import ReactTooltip from "react-tooltip";
+
+import "./Tooltip.scss";
+
+const Tooltip = ({
```
Comment (luwol03) [inline src/Components/Tooltip/Tooltip.jsx:6]: could we add an id to connect the tooltip to the specific wrapper? [see](https://github.com/wwayne/react-tooltip#3-tooltip-not-binding-to-dynamic-content)

File: src/redux/Actions/form.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import { SET_VOCAB_ACTIVE, SET_VOCAB_ACTIVATE } from "./index.js";
+
+export const setVocabActive = ({ active }) => {
+  return {
+    type: SET_VOCAB_ACTIVE,
+    payload: {
+      vocab: {
```
Comment (luwol03) [inline src/redux/Actions/form.js:7]: we dont need this extra object vocab inside of payload

File: src/redux/Reducers/form.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+const initialState = defineState(defaultState)("form");
+
+const formReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case SET_VOCAB_ACTIVE:
+      return {
+        ...state,
+        vocab: {
+          activate: state.vocab.activate,
```
Comment (luwol03) [inline src/redux/Reducers/form.js:20]: could we use `...state.vocab`

File: src/Forms/VocabForm/VocabForm.jsx [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+        setPackages(data);
+      })
+      .catch((err) => {
+        showSnack("error", t("global.fetchError"));
+      });
+  }, [showSnack, t]);
+
+  const onClear = useCallback(() => {
+    setSelectedPackage(null);
+    setSelectedGroup(null);
```
Comment (luwol03) [inline src/Forms/VocabForm/VocabForm.jsx:72]: I think it gives a better UX if we don't delete the package and group when you enter a vocab.

File: src/screens/FirstStartup/FirstStartup.jsx [hunk 14]
```diff
@@ -22,9 +22,8 @@ const FirstStartup = () => {
       open={show}
       onClose={() => setShow(false)}
       renderClose={false}
-      closeOnClickOutside={false}
     >
-      <AddLanguagePackage />
+      <PackageForm />
```
Comment (luwol03) [inline src/screens/FirstStartup/FirstStartup.jsx:26]: on close cb is missing

**Review by luwol03 — APPROVED** [review]

**Pull request merged**